src/providers/rust.rs

```diff
@@ -233,6 +218,141 @@ impl RustProvider {
 
         Ok(false)
     }
+
+    fn resolve_cargo_workspace(app: &App, env: &Environment) -> Result<Option<String>> {
+        if let Some(name) = env.get_config_variable("CARGO_WORKSPACE") {
+            return Ok(Some(name));
+        }
+
+        let manifest = RustProvider::parse_cargo_toml(app)?.context("Missing Cargo.toml")?;
+
+        if let Some(workspace) = manifest.workspace {
+            for default_member in workspace
+                .default_members
+                .iter()
+                .filter(|default_member| !workspace.exclude.contains(default_member))
+            {
+                // a member can have globs
+                if default_member.contains('*') || default_member.contains('?') {
+                    for member in app.find_directories(default_member)? {
+                        let mut manifest =
+                            app.read_toml::<Manifest>(&format!("{}/Cargo.toml", member.display()))?;
+
+                        manifest.complete_from_path(
+                            &app.source.join(format!("{}/Cargo.toml", member.display())),
+                        )?;
+
+                        if let Some(package) = manifest.package {
+                            if !manifest.bin.is_empty() || manifest.lib.is_none() {
+                                return Ok(Some(package.name));
+                            }
+                        }
+                    }
+                } else {
+                    let mut manifest =
+                        app.read_toml::<Manifest>(&format!("{}/Cargo.toml", default_member))?;
+
+                    manifest.complete_from_path(
+                        &app.source.join(format!("{}/Cargo.toml", default_member)),
+                    )?;
+
+                    if let Some(package) = manifest.package {
+                        if !manifest.bin.is_empty() || manifest.lib.is_none() {
+                            return Ok(Some(package.name));
+                        }
```

Comment:
This bit of code is reused a few places. Can it be extracted into a function on the manifest. Something like `manifest.is_empty`

Reply:
We can't extend an implementation of a struct unless we define an extension `trait`, which would be more harder.